crates/red_knot_test/src/parser.rs

```diff
@@ -127,6 +129,118 @@ struct Section<'s> {
 #[newtype_index]
 struct EmbeddedFileId;
 
+#[derive(Debug)]
+pub(crate) struct CodeBlock<'s> {
+    pub(crate) code: &'s str,
+
+    /// The offset of the backticks beginning the code block within the markdown file
+    pub(crate) backtick_offset: TextSize,
+}
+
+/// Holds information about the position and the length of a single code block in a
+/// Markdown file.
+#[derive(Debug)]
+pub(crate) struct CodeBlockDimensions {
+    pub(crate) backtick_offset: TextSize,
+    pub(crate) line_count: usize,
+}
+
+/// Holds information about the position and length of all code blocks that are part of
+/// a single embedded file in a Markdown file. This is used to reconstruct absolute line
+/// numbers (in the Markdown file) from relative line numbers (in the embedded file).
+pub(crate) struct CodeBlockStructure {
+    start_line_and_line_count: Vec<(usize, usize)>,
+}
+
+impl CodeBlockStructure {
+    pub(crate) fn new(
+        md_index: &LineIndex,
+        dimensions: impl Iterator<Item = CodeBlockDimensions>,
+    ) -> CodeBlockStructure {
+        CodeBlockStructure {
+            start_line_and_line_count: dimensions
+                .map(|d| (md_index.line_index(d.backtick_offset).get(), d.line_count))
+                .collect(),
+        }
+    }
+
+    pub(crate) fn to_absolute_line_number(&self, relative_line_number: OneIndexed) -> OneIndexed {
+        // If we have a Markdown section with multiple code blocks like this:
```

Comment:
It might be worth to move this comment into the `CodeBlockStructure`'s doccomment